File: src/train/AlphaZeroTrain.py

import sys
import os
import random
import time
import queue
import traceback
import numpy as np
import torch
import torch.optim as optim
import torch.multiprocessing as mp
from collections import deque

# Add src to path
script_dir = os.path.dirname(os.path.abspath(__file__))
src_dir = os.path.dirname(script_dir)
sys.path.append(src_dir)

from env.BoardEnv import BaseBoardEnv
from agents.AlphaZeroAgent import AlphaZeroAgent
from models.Net import AlphaZeroNet

# --- Configuration ---
N = 3
NUM_WORKERS = 16  # Number of self-play workers
MCTS_SIMS = 400
EPISODES_PER_ITER = 100
BATCH_SIZE = 128
EPOCHS = 10
LR = 0.001
BUFFER_CAPACITY = 50000
NUM_ITERS = 10

# --- Helper Classes ---

class StopTraining(Exception):
    """Raised when training is complete and workers should exit cleanly."""
    pass

class ReplayBuffer:
    def __init__(self, capacity=10000):
        self.buffer = deque(maxlen=capacity)
        
    def push(self, state, policy, value):
        self.buffer.append((state, policy, value))
        
    def sample(self, batch_size):
        return random.sample(self.buffer, batch_size)
    
    def __len__(self):
        return len(self.buffer)

class RemoteModel:
    """
    A wrapper that acts like the model but sends requests to a prediction queue.
    Used by workers to get predictions from the central GPU/inference process.
    """
    def __init__(self, N, request_queue, result_queue, worker_id, stop_event):
        self.dummy_net = AlphaZeroNet(N) # Lightweight, for preprocess only
        self.request_queue = request_queue
        self.result_queue = result_queue
        self.worker_id = worker_id
        self.stop_event = stop_event
        self.device = 'cpu' # Workers run on CPU
        
    def preprocess(self, state):
        # preprocess is a static-like method in AlphaZeroNet
        return self.dummy_net.preprocess(state)
        
    def __call__(self, state_tensor):
        # state_tensor shape: (1, 3, N+1, N+1)
        if self.stop_event.is_set():
            raise StopTraining()
        
        # Send request
        state_tensor.share_memory_()
        self.request_queue.put((self.worker_id, state_tensor))
        
        # Wait for valid response
        while not self.stop_event.is_set():
            try:
                resp = self.result_queue.get(timeout=0.2)
                if isinstance(resp, Exception):
                    raise resp
                logits, value = resp
                return logits, value
            except queue.Empty:
                continue
        
        # If we reach here, stop_event is set — clean exit
        raise StopTraining()
        
    def to(self, device):
        # Ignore device movement, we stay on CPU/Queue
        return self
        
    def eval(self):
        pass

# --- Worker Processes ---

def prediction_worker(N, model_path, request_queue, result_queues, stop_event, device_str):
    """
    Central Inference Server.
    Batches requests from workers and runs model on GPU.
    """
    try:
        # Set threads to 1 to avoid contention with workers and overhead for small batches
        torch.set_num_threads(1)
        device = torch.device(device_str)
        model = AlphaZeroNet(N).to(device)
        
        if os.path.exists(model_path):
            try:
                model.load_state_dict(torch.load(model_path, map_location=device))
                print(f"[Prediction] Loaded model from {model_path}")
            except:
                print(f"[Prediction] Failed to load model, starting fresh.")
        
        model.eval()
        
        while not stop_event.is_set():
            # Collect batch logic
            batch_reqs = []
            
            # Simple adaptive batching
            try:
                # Blocking get for first item
                req = request_queue.get(timeout=0.1)
                batch_reqs.append(req)
            except queue.Empty:
                continue
                
            # Try to get more items to fill batch (without waiting too long)
            # This is the "Batch" part of Batch MCTS
            start_wait = time.time()
            while len(batch_reqs) < 16 and (time.time() - start_wait) < 0.01:
                try:
                    req = request_queue.get_nowait()
                    batch_reqs.append(req)
                except queue.Empty:
                    break
            
            # Process batch
            indices = []
            tensors = []
            reload_cmd = False
            
            for req in batch_reqs:
                worker_id, data = req
                
                # Check for special commands
                if worker_id == 'COMMAND':
                    if data == 'RELOAD':
                        print("[Prediction] Reloading model weights...")
                        try:
                            model.load_state_dict(torch.load(model_path, map_location=device))
                            model.eval()
                            print("[Prediction] Model reloaded.")
                        except Exception as e:
                            print(f"[Prediction] Error reloading model: {e}")
                        reload_cmd = True
                    continue
                
                indices.append(worker_id)
                tensors.append(data)
            
            if not tensors:
                continue
                
            if reload_cmd:
                # If we reloaded, the pending requests might be against old model, 
                # but currently we just process them with new model.
                pass
                
            # Inference
            # Stack: tensors are (1, C, H, W), so stack -> (B, C, H, W)
            # We need to squeeze the dim 0 from each tensor before stack? 
            # AlphaZeroAgent sends .unsqueeze(0).
            # So standard torch.cat is enough
            input_batch = torch.cat(tensors).to(device)
            
            with torch.no_grad():
                logits, values = model(input_batch)
                
            logits = logits.cpu()
            values = values.cpu()
            
            # Distribute results
            for i, worker_id in enumerate(indices):
                # result_queues is a list of queues
                # We return separate tensors (slice them)
                l = logits[i:i+1].clone() # Keep (1, ...) shape, clone to detach from batch storage
                v = values[i:i+1].clone()
                l.share_memory_()
                v.share_memory_()
                result_queues[worker_id].put((l, v))
        
        # --- Graceful shutdown: drain any remaining requests ---
        # Workers may still have pending requests in the queue.
        # Respond with StopTraining so they unblock cleanly.
        _drain_deadline = time.time() + 2.0  # spend at most 2 seconds draining
        while time.time() < _drain_deadline:
            try:
                req = request_queue.get_nowait()
                worker_id, data = req
                if worker_id == 'COMMAND':
                    continue
                # Send an exception so the worker unblocks
                result_queues[worker_id].put(StopTraining())
            except queue.Empty:
                break
                
    except Exception as e:
        print(f"[Prediction] CRASHED: {e}")
        traceback.print_exc()

def self_play_worker(worker_id, N, request_queue, result_queue, game_queue, stop_event, mcts_sims):
    """
    Self-Play Worker.
    Plays games continuously and sends results to game_queue.
    """
    try:
        # Pseudo-random seed for each worker
        np.random.seed(int(time.time()) + worker_id * 1000)
        random.seed(int(time.time()) + worker_id * 1000)
        
        env = BaseBoardEnv(N)
        
        # Initialize Agent with RemoteModel
        remote_model = RemoteModel(N, request_queue, result_queue, worker_id, stop_event)
        agent = AlphaZeroAgent(env, n_simulations=mcts_sims, device='cpu')
        agent.model = remote_model
        
        while not stop_event.is_set():
            env.reset()
            game_history = []
            
            while not env.done:
                # If stop requested inside game, break
                if stop_event.is_set(): break
                
                state_snapshot = agent._clone_state(env)
                
                # Temp schedule
                temp = 1.0 if len(game_history) < 10 else 0.0
                
                action, policy_dict = agent.act(return_probs=True, temperature=temp)
                
                game_history.append([state_snapshot, env.current_player, policy_dict])
                env.step(action)
            
            if stop_event.is_set(): break

            # Game Over
            s1, s2 = env.score
            if s1 > s2: result = 1
            elif s2 > s1: result = -1
            else: result = 0
            
            # Process Game History
            processed_data = []
            for state, player, policy_dict in game_history:
                policy_vec = np.zeros(agent.model.dummy_net.action_size if hasattr(agent.model.dummy_net, 'action_size') else (N+1)*N + N*(N+1), dtype=np.float32)
                # Note: AlphaZeroNet does not expose action_size directly usually, 
                # but we know it for N.
                # Let's fix action_size logic.
                # AlphaZeroTrainer.py used net.action_size? 
                # Checking Net.py: p_fc output size is (N+1)*N + N*(N+1).
                # But Net does not have self.action_size attribute visible in code.
                # AlphaZeroTrainer line 100 used 'net.action_size'.
                # Wait, did I miss that in Net.py?
                # The file I read didn't show self.action_size = ...
                # Let's calculate it manually to be safe.
                action_size = (N+1)*N + N*(N+1)
                
                policy_vec = np.zeros(action_size, dtype=np.float32)
                total_visits = sum(policy_dict.values())
                
                for action, visits in policy_dict.items():
                    idx = agent._action_to_index(action)
                    policy_vec[idx] = visits / total_visits
                    
                z = result if player == 1 else -result
                
                # Optimization: Send numpy array instead of tensor to avoid 
                # "Too many open files" error with shared memory tensors.
                state_numpy = remote_model.preprocess(state).numpy()
                processed_data.append((state_numpy, policy_vec, z))
            
            # Send game data
            game_queue.put(processed_data)
    
    except StopTraining:
        # Clean exit — training is complete
        pass
    except (ConnectionResetError, FileNotFoundError, EOFError, BrokenPipeError):
        # Expected during shutdown — the prediction server closed its end
        if not stop_event.is_set():
            print(f"[Worker {worker_id}] Connection error (unexpected, stop not set)")
    except Exception as e:
        # Only print truly unexpected crashes
        if not stop_event.is_set():
            print(f"[Worker {worker_id}] CRASHED: {e}")
            traceback.print_exc()

# --- Main Trainer ---

def train_alpha_zero_mp():
    mp.set_start_method('spawn', force=True)
    
    models_dir = os.path.join(src_dir, '../models')
    os.makedirs(models_dir, exist_ok=True)
    model_path = os.path.join(models_dir, f"alphazero_{N}x{N}.pth")
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Main Process: Using device: {device}")
    
    # Shared Queues
    # request_queue: (worker_id, data)
    request_queue = mp.Queue(maxsize=1000)
    # result_queues: list of queues, one per worker
    result_queues = [mp.Queue(maxsize=10) for _ in range(NUM_WORKERS)]
    # game_queue: buffer for finished games
    game_queue = mp.Queue(maxsize=100)
    
    stop_event = mp.Event()
    
    # Start Prediction Server
    print("Starting Prediction Server...")
    pred_process = mp.Process(
        target=prediction_worker,
        args=(N, model_path, request_queue, result_queues, stop_event, str(device))
    )
    pred_process.start()
    
    # Start Self-Play Workers
    print(f"Starting {NUM_WORKERS} Self-Play Workers...")
    workers = []
    for i in range(NUM_WORKERS):
        p = mp.Process(
            target=self_play_worker,
            args=(i, N, request_queue, result_queues[i], game_queue, stop_event, MCTS_SIMS)
        )
        p.start()
        workers.append(p)
        
    # Replay Buffer
    replay_buffer = ReplayBuffer(capacity=BUFFER_CAPACITY)
    
    # Load Model for Training (Main process also needs a model instance to train)
    train_net = AlphaZeroNet(N).to(device)
    optimizer = optim.Adam(train_net.parameters(), lr=LR)
    
    if os.path.exists(model_path):
        try:
            train_net.load_state_dict(torch.load(model_path, map_location=device))
            print("Main Process: Loaded existing model.")
        except:
            print("Main Process: Starting fresh model.")
            
    # Main Loop
    try:
        for iteration in range(NUM_ITERS):
            print(f"\n=== Iteration {iteration+1}/{NUM_ITERS} ===")
            start_time = time.time()
            
            # Collection Phase
            games_collected = 0
            print(f"  Collecting {EPISODES_PER_ITER} games...")
            
            while games_collected < EPISODES_PER_ITER:
                try:
                    game_data = game_queue.get(timeout=1.0) # List of steps
                    for step in game_data:
                        state_np, policy, value = step
                        state_tensor = torch.from_numpy(state_np)
                        replay_buffer.push(state_tensor, policy, value)
                    
                    games_collected += 1
                    if games_collected % 10 == 0:
                        print(f"    Collected {games_collected}/{EPISODES_PER_ITER} games")
                except queue.Empty:
                    # Check if processes are alive
                    if not pred_process.is_alive():
                        print("Prediction process died!")
                        stop_event.set()
                        break
                    continue
            
            duration = time.time() - start_time
            print(f"  Collection Time: {duration:.2f}s ({duration/EPISODES_PER_ITER:.2f}s/game)")
            print(f"  Replay Buffer Size: {len(replay_buffer)}")
            
            # Training Phase
            if len(replay_buffer) < BATCH_SIZE:
                print("  Not enough samples to train.")
                continue
            
            print(f"  Training Neural Net...")
            train_net.train()
            total_loss = 0
            
            for _ in range(EPOCHS):
                batch = replay_buffer.sample(BATCH_SIZE)
                states, policies, values = zip(*batch)
                
                states_v = torch.stack(states).to(device)
                policies_v = torch.tensor(np.array(policies), dtype=torch.float32).to(device)
                values_v = torch.tensor(np.array(values), dtype=torch.float32).unsqueeze(1).to(device)
                
                optimizer.zero_grad()
                out_logits, out_value = train_net(states_v)
                
                v_loss = torch.nn.functional.mse_loss(out_value, values_v)
                log_probs = torch.nn.functional.log_softmax(out_logits, dim=1)
                p_loss = -torch.sum(policies_v * log_probs) / BATCH_SIZE
                
                loss = v_loss + p_loss
                loss.backward()
                optimizer.step()
                total_loss += loss.item()
                
            print(f"  Avg Loss: {total_loss / EPOCHS:.4f}")
            
            # Save and Reload
            torch.save(train_net.state_dict(), model_path)
            print(f"  Model saved to {model_path}")
            
            # Trigger Reload in Prediction Server
            request_queue.put(('COMMAND', 'RELOAD'))
            
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        print("Shutting down workers...")
        stop_event.set()
        
        # Give prediction server time to drain pending requests
        pred_process.join(timeout=10)
        if pred_process.is_alive():
            print("  Prediction server didn't stop, terminating...")
            pred_process.terminate()
            pred_process.join(timeout=5)
        
        # Wait for workers with timeout
        for i, p in enumerate(workers):
            p.join(timeout=5)
            if p.is_alive():
                p.terminate()
                p.join(timeout=2)
        
        print("Done.")

if __name__ == '__main__':
    train_alpha_zero_mp()

File: src/cpp/train/AlphaZeroTrain.cpp
#include "AlphaZeroTrain.h"

#include <chrono>
#include <filesystem>
#include <iostream>
#include <fstream>

namespace azb {

AlphaZeroTrainer::AlphaZeroTrainer(const TrainConfig& cfg)
    : cfg_(cfg),
      device_(torch::kCPU),
      replay_buffer_(cfg.buffer_capacity) {

    // ── Build model via ModelHandle factories ──
    if (cfg_.use_patch_net) {
        model_ = make_patch_model(
            cfg.rows, cfg.cols,
            cfg.patch_rows, cfg.patch_cols,
            cfg.local_hidden_size, cfg.local_num_res_blocks,
            cfg.global_hidden_size, cfg.global_num_res_blocks,
            cfg.dropout, cfg.local_model_path);
        std::cout << "[Trainer] Using PatchNet ("
                  << cfg.patch_rows << "x" << cfg.patch_cols
                  << " patches on " << cfg.rows << "x" << cfg.cols
                  << " board)" << std::endl;
    } else {
        model_ = make_standard_model(
            cfg.rows, cfg.cols,
            cfg.hidden_size, cfg.num_res_blocks, cfg.dropout);
        std::cout << "[Trainer] Using AlphaZeroBitNet ("
                  << cfg.hidden_size << "h, "
                  << cfg.num_res_blocks << " blocks)" << std::endl;
    }

    if (torch::cuda::is_available()) {
        device_ = torch::Device(torch::kCUDA);
        std::cout << "[Trainer] Using CUDA" << std::endl;
    } else {
        std::cout << "[Trainer] Using CPU" << std::endl;
    }
    model_.to(device_);

    if (cfg_.resume) {
        const auto path = model_path();
        if (std::filesystem::exists(path)) {
            try {
                model_.load(path);
                model_.to(device_);
                std::cout << "[Trainer] Loaded model from " << path << std::endl;
            } catch (const std::exception& e) {
                std::cerr << "[Trainer] Failed to load model: " << e.what()
                          << ", starting fresh." << std::endl;
            }
        } else {
            std::cerr << "[Trainer] Resume requested but no model found at "
                      << path << ". Starting fresh." << std::endl;
        }
    }
}

std::string AlphaZeroTrainer::model_path() const {
    return cfg_.model_dir + "/alphazero_" +
           std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".pt";
}

void AlphaZeroTrainer::train() {
    std::filesystem::create_directories(cfg_.model_dir);

    const std::string loss_log_path =
        cfg_.model_dir + "/loss_log_" + std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".csv";
    const bool loss_log_exists = std::filesystem::exists(loss_log_path);
    std::ofstream loss_log(loss_log_path, std::ios::app);
    if (!loss_log) {
        std::cerr << "[Trainer] Failed to open loss log file: " << loss_log_path << std::endl;
    } else if (!loss_log_exists) {
        loss_log << "timestamp_ms,phase_idx,phase_name,iter,epoch,epochs,batches,loss,value_loss,policy_loss\n";
    }

    const std::string batch_debug_path =
        cfg_.model_dir + "/batch_debug_" + std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".csv";
    const bool batch_debug_exists = std::filesystem::exists(batch_debug_path);
    std::ofstream batch_debug(batch_debug_path, std::ios::app);
    if (!batch_debug) {
        std::cerr << "[Trainer] Failed to open batch debug file: " << batch_debug_path << std::endl;
    } else if (!batch_debug_exists) {
        batch_debug << "timestamp_ms,phase_idx,phase_name,iter,epoch,epochs,batches,policy_sum_mean,zero_policy_rows,legal_moves_mean\n";
    }

    // Write model architecture sidecar so Python can auto-detect hidden/blocks
    {
        std::ofstream info(cfg_.model_dir + "/model_info.json");
        info << "{\"rows\":" << cfg_.rows
             << ",\"cols\":" << cfg_.cols
             << ",\"hidden_size\":" << cfg_.hidden_size
             << ",\"num_res_blocks\":" << cfg_.num_res_blocks
             << ",\"value_eval\":\"" << azb::value_eval_name(cfg_.value_eval) << "\""
             << ",\"use_patch_net\":" << (cfg_.use_patch_net ? "true" : "false");
        if (cfg_.use_patch_net) {
            info << ",\"patch_rows\":" << cfg_.patch_rows
                 << ",\"patch_cols\":" << cfg_.patch_cols
                 << ",\"local_hidden_size\":" << cfg_.local_hidden_size
                 << ",\"local_num_res_blocks\":" << cfg_.local_num_res_blocks
                 << ",\"global_hidden_size\":" << cfg_.global_hidden_size
                 << ",\"global_num_res_blocks\":" << cfg_.global_num_res_blocks
                 << ",\"local_model_path\":\"" << cfg_.local_model_path << "\"";
        }
        info << "}\n";
    }

    // If no phases provided, create a default one from existing config
    if (cfg_.phases.empty()) {
        Phase p;
        p.name = "Default";
        p.iterations = cfg_.num_iters;
        p.mcts_sims = cfg_.mcts_sims;
        p.episodes_per_iter = cfg_.episodes_per_iter;
        p.epochs = cfg_.epochs;
        p.lr = static_cast<float>(cfg_.learning_rate);
        p.temp_threshold = cfg_.temp_threshold;
        p.temp_explore = cfg_.temp_explore;
        p.temp_exploit = cfg_.temp_exploit;
        cfg_.phases.push_back(p);
    }

    int start_phase_idx = 0;
    int start_iter = 0;
    std::string state_path = cfg_.model_dir + "/train_state_" + std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".txt";
    std::string optim_path = cfg_.model_dir + "/optimizer_state_" + std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".pt";
    std::string replay_path = cfg_.model_dir + "/replay_buffer_" + std::to_string(cfg_.rows) + "x" + std::to_string(cfg_.cols) + ".pt";

    if (cfg_.resume) {
        if (std::filesystem::exists(state_path)) {
            std::ifstream ifs(state_path);
            if (ifs >> start_phase_idx >> start_iter) {
                std::cout << "[Trainer] Resuming from phase " << start_phase_idx << ", iter " << start_iter << std::endl;
            } else {
                std::cerr << "[Trainer] Failed to parse train state. Starting fresh." << std::endl;
                start_phase_idx = 0;
                start_iter = 0;
            }
        } else {
            std::cout << "[Trainer] No train state found to resume. Starting fresh." << std::endl;
        }
    }

    // Persistent optimizer
    torch::optim::Adam optimizer(model_.parameters(),
                                 torch::optim::AdamOptions(cfg_.phases.empty() ? cfg_.learning_rate : cfg_.phases[0].lr));
    if (cfg_.resume && std::filesystem::exists(optim_path)) {
        try {
            torch::load(optimizer, optim_path);
            std::cout << "[Trainer] Loaded optimizer state from " << optim_path << std::endl;
        } catch (const std::exception& e) {
            std::cerr << "[Trainer] Failed to load optimizer state: " << e.what()
                      << ", continuing with fresh optimizer." << std::endl;
        }
    } else if (cfg_.resume) {
        std::cout << "[Trainer] No optimizer state found to resume." << std::endl;
    }

    if (cfg_.resume && std::filesystem::exists(replay_path)) {
        if (replay_buffer_.load(replay_path, cfg_.buffer_capacity)) {
            std::cout << "[Trainer] Loaded replay buffer (" << replay_buffer_.size()
                      << " samples, capacity " << replay_buffer_.capacity() << ") from "
                      << replay_path << std::endl;
        } else {
            std::cerr << "[Trainer] Failed to load replay buffer. Starting fresh." << std::endl;
        }
    } else if (cfg_.resume) {
        std::cout << "[Trainer] No replay buffer found to resume." << std::endl;
    }

    for (size_t p_idx = start_phase_idx; p_idx < cfg_.phases.size(); p_idx++) {
        const auto& phase = cfg_.phases[p_idx];
        std::cout << "\n>>> STARTING PHASE: " << phase.name << " <<<" << std::endl;
        std::cout << "Config: sims=" << phase.mcts_sims 
                  << ", episodes=" << phase.episodes_per_iter
                  << ", lr=" << phase.lr << std::endl;

        // Update learning rate
        for (auto& param_group : optimizer.param_groups()) {
            static_cast<torch::optim::AdamOptions&>(param_group.options()).lr(phase.lr);
        }

        int iter_start_loop = (p_idx == static_cast<size_t>(start_phase_idx)) ? start_iter : 0;

        for (int iter = iter_start_loop; iter < phase.iterations; iter++) {
            std::cout << "\n=== [" << phase.name << "] Iteration " 
                      << (iter + 1) << "/" << phase.iterations << " ===" << std::endl;
            auto iter_start = std::chrono::steady_clock::now();

            // ── Self-Play Phase ──────────────────────────────────────────

            // Create inference model (copy of training model)
            ModelHandle infer_model = model_.clone(cfg_);
            infer_model.to(device_);
            infer_model.eval_mode();

            InferenceServer server(infer_model, device_, cfg_.num_workers,
                                   cfg_.max_inference_batch);

            std::atomic<int> games_collected{0};
            std::atomic<bool> stop_workers{false};

            std::thread server_thread([&server]() { server.run(); });

            std::vector<std::unique_ptr<SelfPlayWorker>> workers;
            std::vector<std::thread> worker_threads;
            for (int w = 0; w < cfg_.num_workers; w++) {
                workers.push_back(std::make_unique<SelfPlayWorker>(
                    w, cfg_, phase, server, replay_buffer_, games_collected, stop_workers));
            }
            for (int w = 0; w < cfg_.num_workers; w++) {
                auto* worker_ptr = workers[w].get();
                worker_threads.emplace_back([worker_ptr]() { worker_ptr->run(); });
            }

            // Wait for enough games
            int last_reported = 0;
            while (games_collected.load() < phase.episodes_per_iter) {
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
                int gc = games_collected.load();
                if (gc >= last_reported + 10) {
                    std::cout << "    Collected " << gc << "/" << phase.episodes_per_iter
                              << " games" << std::endl;
                    last_reported = gc;
                }
            }

            stop_workers.store(true);
            server.stop();
            for (auto& t : worker_threads) if (t.joinable()) t.join();
            if (server_thread.joinable()) server_thread.join();

            auto collect_end = std::chrono::steady_clock::now();
            double collect_secs = std::chrono::duration<double>(collect_end - iter_start).count();
            std::cout << "  Collection: " << collect_secs << "s, Buffer: " << replay_buffer_.size() << std::endl;

            // ── Training Phase ───────────────────────────────────────────

            if (replay_buffer_.size() < cfg_.batch_size) {
                std::cout << "  Wait for more samples..." << std::endl;
                continue;
            }

            std::cout << "  Training..." << std::endl;
            model_.train_mode();
            float total_loss = 0.0f;
            for (int epoch = 0; epoch < phase.epochs; epoch++) {
                const auto stats = train_epoch(optimizer,
                                               batch_debug ? static_cast<std::ostream*>(&batch_debug) : nullptr,
                                               p_idx,
                                               phase.name,
                                               (iter + 1),
                                               (epoch + 1),
                                               phase.epochs);
                total_loss += stats.loss;

                std::cout << "    Epoch " << (epoch + 1) << "/" << phase.epochs
                          << " | loss=" << stats.loss
                          << " (v=" << stats.value_loss
                          << ", p=" << stats.policy_loss
                          << "), batches=" << stats.batches << std::endl;

                if (loss_log) {
                    const auto now = std::chrono::time_point_cast<std::chrono::milliseconds>(
                        std::chrono::system_clock::now());
                    const auto ts_ms = now.time_since_epoch().count();
                    loss_log << ts_ms << ","
                             << p_idx << ","
                             << phase.name << ","
                             << (iter + 1) << ","
                             << (epoch + 1) << ","
                             << phase.epochs << ","
                             << stats.batches << ","
                             << stats.loss << ","
                             << stats.value_loss << ","
                             << stats.policy_loss << "\n";
                    loss_log.flush();
                }
            }
            std::cout << "  Avg Loss: " << total_loss / phase.epochs << std::endl;

            // ── Checkpointing & Adaptation ───────────────────────────────
            
            // Save iteration-specific model
            std::string iter_path = cfg_.model_dir + "/checkpoint_iter" + std::to_string(iter + 1) + ".pt";
            model_.save(iter_path);
            
            // Overwrite latest main model
            model_.save(model_path());
            
            // Register in history
            cfg_.model_history.push_back(iter_path);
            if (static_cast<int>(cfg_.model_history.size()) > cfg_.keep_checkpoints) {
                std::string to_remove = cfg_.model_history.front();
                cfg_.model_history.erase(cfg_.model_history.begin());
                if (std::filesystem::exists(to_remove)) {
                    std::filesystem::remove(to_remove);
                }
            }

            // Adaptive Buffer: GROW capacity
            if (cfg_.buffer_grow > 0) {
                int new_capacity = replay_buffer_.capacity() + cfg_.buffer_grow;
                std::cout << "  Buffer grown to: " << new_capacity << " samples" << std::endl;
                replay_buffer_.set_capacity(new_capacity);
            }

            int saved_iter = iter + 1;
            int saved_phase = p_idx;
            if (saved_iter >= phase.iterations) {
                saved_phase++;
                saved_iter = 0;
            }
            std::ofstream ofs(state_path);
            ofs << saved_phase << " " << saved_iter << "\n";
            ofs.close();
            try {
                torch::save(optimizer, optim_path);
            } catch (const std::exception& e) {
                std::cerr << "[Trainer] Failed to save optimizer state: " << e.what() << std::endl;
            }
            if (!replay_buffer_.save(replay_path)) {
                std::cerr << "[Trainer] Failed to save replay buffer." << std::endl;
            }

            std::cout << "  Models saved to " << iter_path << " and " << model_path() << "\n  State saved." << std::endl;
        }
    }
    std::cout << "\nAll training phases complete." << std::endl;
}

AlphaZeroTrainer::TrainLossStats AlphaZeroTrainer::train_epoch(torch::optim::Adam& optimizer,
                                                               std::ostream* batch_debug_csv,
                                                               size_t phase_idx,
                                                               const std::string& phase_name,
                                                               int iter_1_based,
                                                               int epoch_1_based,
                                                               int epochs_total) {
    const int buf_size = replay_buffer_.size();
    if (buf_size < cfg_.batch_size) return {};

    const int num_batches = std::min(buf_size / cfg_.batch_size, 256);
    float total_loss = 0.0f;
    float total_v_loss = 0.0f;
    float total_p_loss = 0.0f;
    int batches = 0;

    float policy_sum_mean_acc = 0.0f;
    float legal_moves_mean_acc = 0.0f;
    int zero_policy_rows_acc = 0;

    for (int b = 0; b < num_batches; b++) {
        auto batch = replay_buffer_.sample(cfg_.batch_size);
        if (batch.empty()) continue;

        std::vector<torch::Tensor> states, policies, masks;
        std::vector<float> values;
        states.reserve(batch.size());
        policies.reserve(batch.size());
        masks.reserve(batch.size());
        values.reserve(batch.size());

        for (auto& sample : batch) {
            states.push_back(sample.state);
            policies.push_back(sample.policy);
            masks.push_back(sample.legal_mask);
            values.push_back(sample.value);
        }

        auto states_t   = torch::stack(states).to(device_);
        auto policies_t = torch::stack(policies).to(device_);
        auto masks_t    = torch::stack(masks).to(device_);
        auto values_t   = torch::tensor(values, torch::kFloat32).unsqueeze(1).to(device_);

        // Diagnostics: if policy targets are all-zero (or not normalized), policy loss can be exactly 0.
        // Compute lightweight scalar stats per batch and aggregate for the epoch.
        {
            auto policy_row_sums = policies_t.sum(/*dim=*/1);
            policy_sum_mean_acc += policy_row_sums.mean().item<float>();
            zero_policy_rows_acc += policy_row_sums.le(0.0f).sum().item<int>();
            legal_moves_mean_acc += masks_t.sum(/*dim=*/1).mean().item<float>();
        }

        auto [logits, out_value] = model_.forward(states_t);

        auto v_loss = torch::mse_loss(out_value, values_t);

        auto masked_logits = logits + (masks_t - 1.0f) * 1e9f;
        auto log_probs = torch::log_softmax(masked_logits, /*dim=*/1);
        auto p_loss = -(policies_t * log_probs).sum() / static_cast<float>(cfg_.batch_size);

        auto loss = 1.5f * v_loss + p_loss;

        optimizer.zero_grad();
        loss.backward();
        torch::nn::utils::clip_grad_norm_(model_.parameters(), 1.0f);
        optimizer.step();

        total_loss += loss.item<float>();
        total_v_loss += v_loss.item<float>();
        total_p_loss += p_loss.item<float>();
        batches++;
    }

    const int denom = std::max(batches, 1);
    TrainLossStats stats;
    stats.loss = total_loss / denom;
    stats.value_loss = total_v_loss / denom;
    stats.policy_loss = total_p_loss / denom;
    stats.batches = batches;

    if (batch_debug_csv && batches > 0) {
        const float policy_sum_mean = policy_sum_mean_acc / static_cast<float>(batches);
        const float legal_moves_mean = legal_moves_mean_acc / static_cast<float>(batches);

        const auto now = std::chrono::time_point_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now());
        const auto ts_ms = now.time_since_epoch().count();

        (*batch_debug_csv) << ts_ms << ","
                           << phase_idx << ","
                           << phase_name << ","
                           << iter_1_based << ","
                           << epoch_1_based << ","
                           << epochs_total << ","
                           << batches << ","
                           << policy_sum_mean << ","
                           << zero_policy_rows_acc << ","
                           << legal_moves_mean << "\n";
        batch_debug_csv->flush();

        if (zero_policy_rows_acc > 0 || policy_sum_mean < 0.5f) {
            std::cerr << "[Trainer] Warning: policy target stats look off (epoch "
                      << epoch_1_based << "/" << epochs_total
                      << "): policy_sum_mean=" << policy_sum_mean
                      << ", zero_policy_rows=" << zero_policy_rows_acc
                      << ", legal_moves_mean=" << legal_moves_mean << std::endl;
        }
    }

    return stats;
}


}  // namespace azb
